semaphore/semaphore_test.go

```diff
@@ -200,3 +200,23 @@ func TestAllocCancelDoesntStarve(t *testing.T) {
 	}
 	sem.Release(1)
 }
+
+func TestMaxWaiters(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	sem := semaphore.NewWeighted(1, 10)
+	sem.Acquire(ctx, 1)
+
+	for i := 0; i < 10; i++ {
+		go func() {
+			sem.Acquire(ctx, 1)
+			<-ctx.Done()
+		}()
+	}
+
+	time.Sleep(50 * time.Millisecond)
```

Comment:
I'd prefer that each goroutine write to a channel after it acquires the semaphore, and have the test's main thread read 10 values from that channel before attempting to acquire it an 11th time.

Reply:
Oooh right of course. Hmm. Well.

Writing to a channel _before_ trying to acquire the semaphore would _probably_ solve the problem: imo it's much more likely that the failure mode here is "not all goroutines started prior to the main goroutine trying to acquire" than "all goroutines started but some got pre-empted prior to their acquire call". I think that this approach is approximately as likely to fail as sleeping (it didn't fail once in 1k runs on my local machine), and ~50ms faster.

Reply:
I tried this:

```go
func TestMaxWaiters(t *testing.T) {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	sem := semaphore.NewWeighted(1, 10)
	sem.Acquire(ctx, 1)

	var wg sync.WaitGroup
	wg.Add(10)
	for i := 0; i < 10; i++ {
		go func() {
			wg.Done()
			sem.Acquire(ctx, 1)
			<-ctx.Done()
		}()
	}

	wg.Wait()
	err := sem.Acquire(ctx, 1)
	if err != semaphore.ErrMaxWaiters {
		t.Errorf("expected error when maxWaiters was reached, but got %#v", err)
	}
}
```

And it hung approximately one in five runs.

But doing it with the channel seems to succeed pretty reliably even though it also has a race condition. Interesting! I pushed the channel version.

Reply:
Woah, I would expect this waitgroup version to have basically the same behavior as the channel version. But I guess the function call to wg.Done() is a lot more likely to be selected as a yield point than just writing to a channel.